Fuel cell bug fixes

EnergyArchmage opened the pull request:
Pull request overview
---------------------
collection of fixes related to fuel cell model.
address #6051 , #6609 , and #6347 


### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

EnergyArchmage committed:
correct plant type of num for fuel cell exhaust HX

EnergyArchmage committed:
correct child object name for FC exhaust HX

EnergyArchmage committed:
elevate warnings for missing ElecticLoadCenter objects  to fatal

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into fuel-cell-thermal-fixes

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@EnergyArchmage @lgentile it has been 14 days since this pull request was last updated.

rraustad commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -1974,30 +1985,39 @@ GeneratorController::GeneratorController(std::string const &objectName,
     } else if (UtilityRoutines::SameString(objectType, "Generator:PVWatts")) {
         generatorType = GeneratorType::pvWatts;
         compGenTypeOf_Num = DataGlobalConstants::iGeneratorPVWatts;
         compPlantTypeOf_Num = DataPlant::TypeOf_Other;
```
Since Generator:Photovoltaic above set compPlantName, why doesn't this object, Generator:PVWatts, need to set compPlantName? Same question for Generator:WindTurbine below. Is compPlantName only used somehow for heat reclaim?

EnergyArchmage replied:
those types of generators don't have thermal generation. Only "CHP" types would be connected to plant

EnergyArchmage committed:
Merge branch 'develop' into fuel-cell-thermal-fixes

EnergyArchmage committed:
Set plant operation scheme type to demand only if CHP generator on demand side

Myoldmopar commented:
@EnergyArchmage at this point, the diffs look to be both related to the fixes here, and also unrelated due to diffs from develop.  Do you feel like your fixes are completely done sans unit test?  If so I'll go ahead and review it and test it assuming your pushing on unit tests and we'll move forward and see what happens.

EnergyArchmage commented:
The fixes in here are done and are helpful.  I am working on a new integration test file that runs the fuel cell model in FollowThermal control mode.  But having a hard time getting the new file to run well so it will have to wait.

EnergyArchmage committed:
more fuel cell fixes

add trap for negative log, allow not locating in a zone for internal heat gains, fix arguments in call to plant

Myoldmopar commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -207,20 +207,24 @@ void ElectricPowerServiceManager::getPowerManagerInput()
         }
     } else {
         // issue #4639. see if there are any generators, inverters, converters, or storage devcies, that really need a ElectricLoadCenter:Distribution
+        bool errorsFound(false);
```
Simple change in here, warnings to severe followed by a fatal.  OK.

Myoldmopar commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -1924,7 +1934,7 @@ Real64 ElectPowerLoadCenter::calcLoadCenterThermalLoad()
         for (auto &g : elecGenCntrlObj) {
             plantNotFound = false;
-            PlantUtilities::ScanPlantLoopsForObject(g->name,
+            PlantUtilities::ScanPlantLoopsForObject(g->compPlantName,
```
Subtle.  OK, so this member variable is going to be set somewhere pretty early on in the code then.

Myoldmopar commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -1938,6 +1948,7 @@ Real64 ElectPowerLoadCenter::calcLoadCenterThermalLoad()
                                                     plantNotFound);
             if (!plantNotFound) g->plantInfoFound = true;
         }
+        myCoGenSetupFlag_ = false;
```
I was expecting to see this as a new member variable, but I guess it's already in there.  In which case this should've been set to false when it was first added.  OK.

Myoldmopar commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -1974,30 +1985,39 @@ GeneratorController::GeneratorController(std::string const &objectName,
         generatorType = GeneratorType::fuelCell;
         compGenTypeOf_Num = DataGlobalConstants::iGeneratorFuelCell;
-        compPlantTypeOf_Num = DataPlant::TypeOf_Generator_FCStackCooler;
+        // fuel cell has two possible plant component types, stack cooler and exhaust gas HX.
```
Setting this up right here in the constructor, good.  I was trying to see where compPlantName was getting assigned to the member variable, but that's implied here.

Myoldmopar commented:
OK, the changes look good.  I'm going to merge this in unless I hear a reason to not do it.
